lib/WebService/Async/CustomerIO.pm

```diff
@@ -0,0 +1,266 @@
+
+use constant {
+    TRACKING_END_POINT => 'https://track.customer.io/api/v1',
+    API_END_POINT      => 'https://api.customer.io/v1/api',
+    RPS_LIMIT_TRACKING => 30,
+    RPS_LIMIT_API      => 10,
+};
+
+=head2 new(%params) -> obj
+
+Creates a new api client object
+
+Parameters:
+- site_id
+- api_key
+
+=cut
+
+sub configure {
+    my ($self, %args) = @_;
+    for my $k (qw(site_id api_key)) {
+        Carp::croak "Missing requeread argument: $k" unless exists $args{$k};
+        $self->{$k} = delete $args{$k} if exists $args{$k};
+    }
+
+    $self->{tracking_ratelimiter} =
+        WebService::Async::CustomerIO::RateLimiter->new(
+            limit    => RPS_LIMIT_TRACKING,
+            interval => 1,
+        );
+    $self->add_child($self->{tracking_ratelimiter});
+
+    $self->{api_ratelimiter} =
+        WebService::Async::CustomerIO::RateLimiter->new(
+            limit    => RPS_LIMIT_API,
+            interval => 1,
+        );
+    $self->add_child($self->{api_ratelimiter});
+
+    $self->SUPER::configure(%args);
+}
+
+=head2 site_id
+
+=cut
+
+sub site_id {shift->{site_id}}
+
+=head2 api_key
+
+=cut
+
+sub api_key {shift->{api_key}}
+
+=head2 api_ratelimiter
+
+Getter returns RateLimmiter for regular API endpoint.
+
+=cut
+
+sub api_ratelimiter {shift->{api_ratelimiter}}
+
+=head2 tracking_ratelimiter
+
+Getter returns RateLimmiter for tracking API endpoint.
+
+=cut
+
+sub tracking_ratelimiter {shift->{tracking_ratelimiter}}
+
+=head2 tracking_request($method, $uri, $data) -> future($data)
+
+Sending request to tracking API end point.
+
+=cut
+
+sub tracking_request {
+    my ($self, $method, $uri, $data) = @_;
+    return $self->tracking_ratelimiter->acquire->then(sub {
+        $self->_request($method, join(q{/} => (TRACKING_END_POINT, $uri)), $data);
+    });
+}
+
+=head2 api_request($method, $uri, $data) -> future($data)
+
+Sending request to regular API end point.
+
+=cut
+
+sub api_request {
+    my ($self, $method, $uri, $data) = @_;
+
+    return $self->api_ratelimiter->acquire->then(sub {
+        $self->_request($method, join(q{/} => (API_END_POINT, $uri)), $data);
+    });
+}
+
+sub _request {
+    my ($self, $method, $uri, $data) = @_;
+
+    my $body = encode_json_utf8($data) if $data;
+    $body //= '' if $method eq 'POST';
+
+    return $self->_ua->do_request(
+            method  => $method,
+            uri     => $uri,
+            user    => $self->site_id,
+            pass    => $self->api_key,
+            !defined $body ? () : (
+                content      => $body,
+                content_type => 'application/json',
+            ),
+    )->catch(sub {
+        my ($code_msg, $err_type, $response) = @_;
+        return Future->fail(@_) unless $err_type && $err_type eq 'http';
+
+        my $code = $response->code;
+        my $request_data = {method => $method, uri => $uri, data => $data};
+
+        return Future->fail('RESOURCE_NOT_FOUND', $request_data)  if $code == 404;
```

Comment:
For failures, note that it's a 3-part system: the exception message or object, the category, then any details. I'd suggest `Future->fail('RESOURCE_NOT_FOUND', 'customer.io', request => $request_data)` here for example.

Reply:
Fixed.